### ISSUE TITLE
NAT model training get different results in different versions of fairseq

### ISSUE THREAD

**Issue by haorannlp** [pytorch/fairseq#3760]
### Before asking:   
1. search the issues.   
2. search the docs.    

#### What is your question?
I cloned the latest version of `fairseq 1.0.0a0+c1624b2 `  and trained a NAT model `CMLM` on `WMT'14 en-de distill`. The best validation loss I got is `7.407`. However, my previous experiments (same experiments, but with different fairseq version `1.0.0a0+1c0439b `) show that  the best validation loss is `3.328`. I'm not sure if this is due to the version of fairseq. 

Here I list the packages of the environments in the above two experiments:
```
Package                Version             Location
---------------------- ------------------- ----------------------------------------------
antlr4-python3-runtime 4.8
certifi                2021.5.30
cffi                   1.14.6
Cython                 0.29.24
fairseq                1.0.0a0+c1624b2    
hydra-core             1.0.7
importlib-resources    5.2.0
mkl-fft                1.3.0
mkl-random             1.2.2
mkl-service            2.4.0
numpy                  1.21.0
olefile                0.46
omegaconf              2.0.6
Pillow                 8.3.1
pip                    21.1.3
portalocker            2.0.0
pycparser              2.20
PyYAML                 5.4.1
regex                  2021.7.6
sacrebleu              1.5.1
setuptools             52.0.0.post20210125
six                    1.16.0
torch                  1.6.0
torchvision            0.7.0
tqdm                   4.61.2
typing-extensions      3.10.0.0
wheel                  0.36.2
zipp                   3.5.0
```
```
Package                Version             Location
---------------------- ------------------- ------------------------------------------
antlr4-python3-runtime 4.8
certifi                2020.12.5
cffi                   1.14.5
click                  7.1.2
Cython                 0.29.22
fairseq                1.0.0a0+1c0439b     
future                 0.18.2
hydra-core             1.0.6
importlib-resources    5.1.1
joblib                 1.0.1
mkl-fft                1.3.0
mkl-random             1.1.1
mkl-service            2.3.0
numpy                  1.20.0
olefile                0.46
omegaconf              2.0.6
Pillow                 8.1.0
pip                    21.0.1
portalocker            2.2.1
pycparser              2.20
PyYAML                 5.4.1
regex                  2020.11.13
sacrebleu              1.5.0
sacremoses             0.0.44
```

The `wmt'14 en-de distillation` data I was using is from `https://github.com/pytorch/fairseq/blob/master/examples/nonautoregressive_translation/README.md`.

The training command is 
```
python train.py data-bin/wmt14_en_de_distill  --save-dir ./checkpoints/cmlm_300k \
    --ddp-backend=legacy_ddp \
    --task translation_lev \
    --criterion nat_loss \
    --arch cmlm_transformer \
    --noise random_mask \
    --share-all-embeddings \
    --optimizer adam --adam-betas '(0.9,0.999)' \
    --lr 0.0005 --lr-scheduler inverse_sqrt \
    --stop-min-lr '1e-09' --warmup-updates 10000 \
    --warmup-init-lr '1e-07' --label-smoothing 0.1 \
    --dropout 0.3 --weight-decay 0.01 \
    --decoder-learned-pos \
    --encoder-learned-pos \
    --apply-bert-init \
    --log-format 'simple' --log-interval 100 \
    --fixed-validation-seed 7 \
    --max-tokens 8000 \
    --update-freq 4 \
    --save-interval 1 \
    --fp16 \
    --keep-last-epochs 80 \
--max-update 300000

```

I've checked the `translation.py`, `translation_lev.py`, `nat_loss.py`, `cmlm_transformer.py`, `nonautoregressive_transformer.py` and I did not find anything strange.

If you have extra computation resources, could you do me a favor by running the above training command with the same fairseq version `fairseq 1.0.0a0+c1624b2`? (mine was running on 4 TESLA-V100)

#### What's your environment?

 - fairseq Version (e.g., 1.0 or master): `fairseq 1.0.0a0+c1624b2 `
 - PyTorch Version (e.g., 1.0): 1.6
 - OS (e.g., Linux): Linux
 - How you installed fairseq (`pip`, source): source
 - Build command you used (if compiling from source):
 - Python version: 3.7
 - CUDA/cuDNN version: 11.0
 - GPU models and configuration:
 - Any other relevant information:

**Comment by haorannlp** [comment]
It's the data problem

**Comment by arvieFrydenlund** [comment]
@haorannlp can you expand on the data problem?  I think I might be having a similar issue.
